### PR TITLE
Variance and covariance layers

### PR DESCRIPTION
**Pull request by timmoon10**
This PR implements layers to compute the variance within a mini-batch sample and the covariance between two inputs. These are implemented for both CPU and GPU and for both data-parallel and model-parallel data. We provide the option to compute either biased or unbiased estimates (i.e. sample or population statistics). Gradient checks on all variations of these layers come back clean.

My main motivation for this PR is to reproduce functionality in the deprecated metrics (see #193), namely Pearson correlation and R2. We can now, in principle, implement these metrics by stitching together simpler layers. This is what TensorFlow does and I think it is much more robust and maintainable than trying to implement monolithic compute kernels. This workflow might be a pain if you are manually writing prototext model files, but that is another argument for moving toward using Python model scripts or the Keras frontend.

### CODE REVIEW

**Comment by bvanessen** [comment]
How hard would it be to implement a compound layer that uses primitive layers internally but presents itself as a single prototext layer. We have discussed this before.


Brian C. Van Essen

vanessen1@llnl.gov<mailto:vanessen1@llnl.gov>

(w) 925-422-9300

(c) 925-290-5470

---

Sent from my iPhone

On Oct 17, 2018, at 6:37 PM, Tim Moon <notifications@github.com<mailto:notifications@github.com>> wrote:


This PR implements layers to compute the variance within a mini-batch sample and the covariance between two inputs. These are implemented for both CPU and GPU and for both data-parallel and model-parallel data. We provide the option to compute either biased or unbiased estimates (i.e. sample or population statistics). Gradient checks on all variations of these layers come back clean.

My main motivation for this PR is to reproduce functionality in the deprecated metrics (see #193<https://github.com/LLNL/lbann/issues/193>), namely Pearson correlation and R2. We can now, in principle, implement these metrics by stitching together simpler layers. This is what TensorFlow does and I think it is much more robust and maintainable than trying to implement monolithic compute kernels. It is a pain if you are manually writing prototext model files, but that is another argument for moving toward Python model scripts or the Keras frontend.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/LLNL/lbann/pull/660

Commit Summary

  *   Variance layer.
  *   Covariance layer.
  *   Bamboo unit tests for variance and covariance layers.
  *   Merge branch 'develop' into feature-variance-layer

File Changes

  *   A bamboo/unit_tests/test_unit_layer_covariance.py<https://github.com/LLNL/lbann/pull/660/files#diff-0> (41)
  *   R bamboo/unit_tests/test_unit_layer_l2_norm2.py<https://github.com/LLNL/lbann/pull/660/files#diff-1> (30)
  *   A bamboo/unit_tests/test_unit_layer_variance.py<https://github.com/LLNL/lbann/pull/660/files#diff-2> (41)
  *   M include/lbann/layers/CMakeLists.txt<https://github.com/LLNL/lbann/pull/660/files#diff-3> (1)
  *   A include/lbann/layers/misc/CMakeLists.txt<https://github.com/LLNL/lbann/pull/660/files#diff-4> (7)
  *   A include/lbann/layers/misc/covariance.hpp<https://github.com/LLNL/lbann/pull/660/files#diff-5> (127)
  *   A include/lbann/layers/misc/variance.hpp<https://github.com/LLNL/lbann/pull/660/files#diff-6> (106)
  *   M include/lbann/lbann.hpp<https://github.com/LLNL/lbann/pull/660/files#diff-7> (4)
  *   A model_zoo/tests/layer_tests/model_covariance.prototext<https://github.com/LLNL/lbann/pull/660/files#diff-8> (110)
  *   R model_zoo/tests/layer_tests/model_l2_norm2.prototext<https://github.com/LLNL/lbann/pull/660/files#diff-9> (0)
  *   A model_zoo/tests/layer_tests/model_variance.prototext<https://github.com/LLNL/lbann/pull/660/files#diff-10> (96)
  *   M src/layers/CMakeLists.txt<https://github.com/LLNL/lbann/pull/660/files#diff-11> (1)
  *   A src/layers/misc/CMakeLists.txt<https://github.com/LLNL/lbann/pull/660/files#diff-12> (17)
  *   A src/layers/misc/covariance.cpp<https://github.com/LLNL/lbann/pull/660/files#diff-13> (190)
  *   A src/layers/misc/covariance.cu<https://github.com/LLNL/lbann/pull/660/files#diff-14> (337)
  *   A src/layers/misc/variance.cpp<https://github.com/LLNL/lbann/pull/660/files#diff-15> (170)
  *   A src/layers/misc/variance.cu<https://github.com/LLNL/lbann/pull/660/files#diff-16> (245)
  *   M src/proto/factories/layer_factory.cpp<https://github.com/LLNL/lbann/pull/660/files#diff-17> (11)
  *   M src/proto/lbann.proto<https://github.com/LLNL/lbann/pull/660/files#diff-18> (16)

Patch Links:

  *   https://github.com/LLNL/lbann/pull/660.patch
  *   https://github.com/LLNL/lbann/pull/660.diff

**Comment by timmoon10** [comment]
It'll take a big refactor of the layer and model classes and it'll probably result in less robust code. We've developed the model class to manage the layer class, and thus they have all sorts of subtle interactions. If we want to implement a compound layer that manages several subordinate layers, it will need to reimplement a good chunk of the functionality in the base layer class and model class.

I think we would benefit a lot if we developed a dedicated front-end, preferably in Python. Our C++-based back-end could just take in a DAG of basic compute kernels and just be focused on chugging through it as efficiently as possible. Simplifying the expectations for the back-end would also allow us to make the code simpler and more robust. By moving model construction into Python, it would be easier to prototype complicated structures (e.g. LSTMS, Pearson correlation) and intricate control flow (e.g. GANs and multi-modal training) without needing to dive into our C++ and think about the implications for the layer class or model class.

**Comment by samadejacobs** [comment]
I think having building block as implemented in this PR is good but it will also be good to provide API for common metrics. TensorFlow has a very long list of metrics [here](https://www.tensorflow.org/api_docs/python/tf/contrib/metrics), we could provide them on demand. One test for this PR will be that we can reproduce Pearson correlation/R2 in models where we have used them.

**Comment by timmoon10** [comment]
Those metrics are mostly implement in Python.
